californium-core/src/main/java/org/eclipse/californium/core/network/serialization/UdpDataSerializer.java

```diff
@@ -0,0 +1,142 @@
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Matthias Kovatsch - creator and main architect
+ *    Martin Lanter - architect and re-implementation
+ *    Dominique Im Obersteg - parsers and initial implementation
+ *    Daniel Pauli - parsers and initial implementation
+ *    Kai Hudalla - logging
+ *    Bosch Software Innovations GmbH - turn into utility class with static methods only
+ *    Joe Magerramov (Amazon AWS) - CoAP over TCP support
+ ******************************************************************************/
+package org.eclipse.californium.core.network.serialization;
+
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.CODE_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.MESSAGE_ID_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.OPTION_DELTA_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.OPTION_LENGTH_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.PAYLOAD_MARKER;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.TOKEN_LENGTH_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.TYPE_BITS;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.VERSION;
+import static org.eclipse.californium.core.coap.CoAP.MessageFormat.VERSION_BITS;
+
+import java.util.List;
+
+import org.eclipse.californium.core.coap.CoAP.Code;
+import org.eclipse.californium.core.coap.Message;
+import org.eclipse.californium.core.coap.Option;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+
+/**
+ * The DataSerialized serializes outgoing messages to byte arrays.
+ */
+public final class UdpDataSerializer implements DataSerializer {
+
+	@Override
+	public byte[] serializeRequest(Request request) {
+		DatagramWriter writer = new DatagramWriter();
+		Code code = request.getCode();
+		serializeMessage(writer, request, code == null ? 0 : code.value);
+		return writer.toByteArray();
+	}
+
+	@Override
+	public byte[] serializeResponse(Response response) {
+		DatagramWriter writer = new DatagramWriter();
+		serializeMessage(writer, response, response.getCode().value);
+		return writer.toByteArray();
+	}
+
+	@Override
+	public byte[] serializeEmptyMessage(Message message) {
+		DatagramWriter writer = new DatagramWriter();
+		serializeMessage(writer, message, 0);
+		return writer.toByteArray();
+	}
+
+	private void serializeMessage(final DatagramWriter writer, final Message message, final int code) {
+		if (message.getToken() == null)
+			throw new NullPointerException("No Token has been set, not even an empty byte[0]");
```

Comment:
Can you take another look. The tests are passing for me now - I think using
TCPMatcher instead of UDP was the root cause of all errors. TCPMatcher
wasn't setting token on empty messages (since it never expects to see
them). I still added one for good measure.

I haven't fixed signed-off footer yet. If PR looks good otherwise I can
repost a new PR with correct signed-by footers.



On Tue, Jun 28, 2016 at 11:50 PM Kai Hudalla <notifications@github.com>
wrote:

> In
> californium-core/src/main/java/org/eclipse/californium/core/network/serialization/UdpDataSerializer.java
> <https://github.com/eclipse/californium/pull/61#discussion_r68893703>:
>
> > +	public byte[] serializeResponse(Response response) {
> > +		DatagramWriter writer = new DatagramWriter();
> > +		serializeMessage(writer, response, response.getCode().value);
> > +		return writer.toByteArray();
> > +	}
> > +
> > +	@Override
> > +	public byte[] serializeEmptyMessage(Message message) {
> > +		DatagramWriter writer = new DatagramWriter();
> > +		serializeMessage(writer, message, 0);
> > +		return writer.toByteArray();
> > +	}
> > +
> > +	private void serializeMessage(final DatagramWriter writer, final Message message, final int code) {
> > +		if (message.getToken() == null)
> > +			throw new NullPointerException("No Token has been set, not even an empty byte[0]");
>
> Well, the null token itself is not a violation itself, I guess, since it
> still results in a zero length token being serialized on the wire (in
> current Californium). However, we might be violating the requirement to
> echo the request's token in ACK and RST messages ...
>
> Server/Client split: I like the idea of having a "client only" and a
> "server that can do both" connector :-) But we can probably introduce that
> at a later point and for now stick with the existing client and server.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/californium/pull/61/files/6554875f60f6f9b9c6b35ed2a14fc1f1a6c1d5bf#r68893703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APu4MgFSSr4LW8Fu4uJdRqGVnY88RaHuks5qQhWXgaJpZM4I9Q-g>
> .
>